### PR TITLE
Support generating field metadata for some runtime introspection

### LINKED ISSUE #79: Proposal: Expose shift and mask values for fields
I've got code that does a read-modify-write by generating a value and a mask, and sending those two values off to another function that actually does the access. It would be useful to be able to get the shift and mask values out of a generated struct definition. I'm imagining something like

```rust
use bitbybit::{bitfield, mask, shift};

#[bitfield(u32)]
struct GICD_TYPER {
    #[bits(11..=15, r)]
    lspi: u5,

    #[bit(10, r)]
    security_extn: bool,

    #[bits(5..=7, r)]
    cpu_number: u3,

    #[bits(0..=4, r)]
    itlines_number: u5,
}

assert_eq!(mask!(GICD_TYPER::cpu_number), 0x7u32)
assert_eq!(shift!(GICD_TYPER::cpu_number), 5usize)
```

where `#[bitfield]` generates constants for the values it's creating (with some name mangling to avoid collisions), and `mask!`/`shift!` take care of matching that mangled name.

Is this a feature that would be likely to be accepted as a PR? If so, I might take a swing at it.

### PR DESCRIPTION
**Pull request by zackyancey**
resolves #79 

Here's what it looks like:

```rs
#[bitfield(u32, introspect)]
struct Bitfield {
    #[bits(0..=7)]
    a: u8,
    #[bits(8..=15)]
    b: [u8; 2],
    #[bits([24..=25, 30..=31])]
    c: u4,
    #[bit(29)]
    d: bool,
}

// <NAME>_BITS exposes the "bits" value(s):
assert_eq!(Bitfield::A_BITS, 0..=7);
assert_eq!(Bitfield::B_BITS, 8..=15);
assert_eq!(Bitfield::C_BITS, [24..=25, 30..=31]);
assert_eq!(Bitfield::D_BITS, 29..=29);

// Arrays also have <NAME>_COUNT and <NAME>_STRIDE
assert_eq!(Bitfield::B_COUNT, 2);
assert_eq!(Bitfield::B_STRIDE, 8);

// <name>_mask() returns a mask for the field
assert_eq!(Bitfield::a_mask(), 0x000000FF);
assert_eq!(Bitfield::b_mask(0), 0x0000FF00);
assert_eq!(Bitfield::b_mask(1), 0x00FF0000);
assert_eq!(Bitfield::c_mask(), 0xC3000000);
assert_eq!(Bitfield::d_mask(), 0x20000000);
```

A couple design decisions that may be worth discussing:

* `*_BITS` is a single range for contiguous bitfields, and a slice of ranges for non-contiguous ones. It could be simpler to always generate a slice (eg. `A_BITS == [0..=7]`), but since most fields are contiguous and the single range is much more ergonomic in that case, I went with that. I'm a bit on the fence about that decision though.
* `*_mask()` is a const method, so that we can accomodate arrays. The main alternatives would be:
  * Have a single mask constant, and the user can shift based on `*_COUNT` and `*_STRIDE`. That would work, but since we've already generated 90% of the work, may as well give them a function.
  * Generate mask as a slice (eg `const B_MASK = [0xFF00, [0xFF0000]`). That could be a lot of constants for large arrays though.
* These values can be enabled per-struct with an attribute, or globally with a feature. Having it as a feature allows users to turn it on by default, and also to get introspection data from external defintions that may not have included the attribute.
* I'm open to changing the `introspect` name, as I think introspection implies a bit more than what we're generating here. Not really sure what else to go with though.

### CODE REVIEW

**Comment by danlehmann** [comment]
This looks really great. I have a minor suggestion for adding a test, but otherwise I'd like to take this as-is

**Comment by zackyancey** [comment]
Added a changelog entry. What's your test suggestion? The one test I added to `bitfield_tests.rs` is pretty basic, so if there are any edge cases you think are worth adding, I can add them.

File: bitbybit/src/bitfield/mod.rs [hunk 1]
```diff
@@ -359,6 +372,40 @@ fn setter_name(field_name: &Ident) -> Ident {
...
+
+    syn::parse_str::<Ident>(&format!("{field_name_without_prefix}_mask"))
+        .unwrap_or_else(|_| panic!("bitfield!: Error creating mask name"))
+}
+
+fn const_name(field_name: &Ident, suffix: &str) -> Ident {
+    // The field might have started with r#. If so, it was likely used for a keyword. This can be dropped here
+    let field_name_without_prefix = {
+        let s = field_name.to_string();
+        if let Some(s) = s.strip_prefix("r#") {
```
[the remark this answers is not in the document]
Comment (zackyancey) [reply bitbybit/src/bitfield/mod.rs:394]: Good call, added both.

**Comment by danlehmann** [comment]
Thanks for your contribution!

**Review by danlehmann — APPROVED** [review]
